shell/browser/api/electron_api_download_item.cc

```diff
@@ -71,65 +97,89 @@ DownloadItem::~DownloadItem() {
     download_item_->RemoveObserver(this);
     download_item_->Remove();
   }
+}
 
-  // Remove from the global map.
-  g_download_item_objects.erase(weak_map_id());
+bool DownloadItem::CheckDestroyed() const {
+  if (!download_item_) {
+    gin_helper::ErrorThrower(v8::Isolate::GetCurrent())
+        .ThrowError("Object has been destroyed");
+    return false;
+  }
+  return true;
 }
 
 void DownloadItem::OnDownloadUpdated(download::DownloadItem* item) {
+  if (!CheckDestroyed())
+    return;
   if (download_item_->IsDone()) {
     Emit("done", item->GetState());
-    // Destroy the item once item is downloaded.
-    base::ThreadTaskRunnerHandle::Get()->PostTask(FROM_HERE,
-                                                  GetDestroyClosure());
+    Unpin();
```

Comment:
Is this guaranteed to always be reached s.t. the `Pin();` call is always balanced out?

Reply:
Hm good question! The previous logic was that the V8 object would be kept alive through the `g_download_item_objects` map until the destructor of the C++ object was called, which was only ever triggered by this `GetDestroyClosure` call here. But it might have been leaky before. I'll look into it. I'm fairly confident at least that this doesn't introduce a _new_ memory leak.

Reply:
In `download_item.h`:

```
  // Returns true if the download is in a terminal state. This includes
  // completed downloads, cancelled downloads, and interrupted downloads that
  // can't be resumed.
  virtual bool IsDone() const = 0;
```

So I think between this and the `Unpin()` I added after this comment (in `OnDownloadDestroyed`) we should be good.

I could imagine an argument that we should Unpin when a download is paused and Pin again when it's resumed (since if you pause a download and then lose the reference, you're never going to resume it), but I think this is OK as it is.